pkg/client/unversioned/clientcmd/merged_client_builder.go

```diff
@@ -112,8 +112,11 @@ func (config *DeferredLoadingClientConfig) ClientConfig() (*restclient.Config, e
 		//   "empty due to defaults"
 		// TODO: this shouldn't be a global - the client config rules should be
 		//   handling this.
-		defaultConfig, err := DefaultClientConfig.ClientConfig()
-		if err == nil && !reflect.DeepEqual(mergedConfig, defaultConfig) {
+		defaultConfig, defErr := DefaultClientConfig.ClientConfig()
+		if IsConfigurationInvalid(defErr) && !IsEmptyConfig(err) {
```

Comment:
@smarterclayton 

> Why are you checking IsEmptyConfig again then? It's checked 3 lines above.

That check doesn't do what this check is doing. If err is `EmptyConfigError`

```go
case err != nil && !IsEmptyConfig(err):
```

falls through, which means code enters 

```go
case mergedConfig != nil:
```
when `mergedConfig` is not `nil`. But `mergedConfig` can still be non-nil and empty. I don't know if that ever happens in practice. But there is a unit test for this. See - https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/clientcmd/merged_client_builder_test.go#L141.

>> A configuration that is both default and non-empty must still be checked. DefaultClientConfigs are only sometimes empty.

> Yes - if you do this you break kubectl's behavior here, which does have a default.

I am not sure what "this" is here. The context/quote isn't helping either. What breaks kubectl's behavior?

> I see the moved deep equal - so if you remove the IsEmptyConfig check we're back to the original PR + test and I'm ok with it.

The original PR doesn't work. See a broken test from the original PR here - https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/32438/kubernetes-pull-test-unit-integration/44167/

The original PR doesn't check for the case when `mergedConfig` is empty but not nil.